server/controllers/userAuth.js

```diff
@@ -130,11 +130,84 @@ const signUp = (req, resp, next) => {
     });
 };
 
-const confirmEmail = (req, resp) => {};
+const confirmEmail = (req, resp) => {
+  const { hash: signUpHash } = req.params;
+
+  User.findOneAndUpdate(
+    {
+      signUpHash,
+      signUpHashExpirationDate: { $gte: new Date() }
+    },
+    {
+      activatedAt: new Date(),
+      isActive: true,
+      signUpHash: null,
+      signUpHashExpirationDate: null
+    }
+  )
+    .lean()
+    .exec()
+    .then(user => {
+      if (!user) {
+        throw new Error();
+      }
+      resp.redirect('/sign-up/success');
+    })
+    .catch(() => {
+      resp.redirect(`/sign-up/failed/${signUpHash}`);
+    });
+};
+
+const resendSignUpConfirmationLink = (req, resp, next) => {
+  const { hash } = req.body;
+  const sanitizedHash = sanitize(hash);
+
+  User.findOne({
+    signUpHash: sanitizedHash,
+    isActive: false
+  })
+    .lean()
+    .exec()
+    .then(user => {
+      if (!user) {
+        throw new Error();
+      }
+
+      const { _id, displayName, email } = user;
+      const signUpHash = crypto.randomBytes(32).toString('hex');
+      const expirationDate = new Date().getTime() + 3600000;
+
+      return User.findOneAndUpdate(
```

Comment:
do you need to find this user again - you already have it in `user` doc - just update it and save in place.